core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -52,6 +59,175 @@ static TEE_Result ffa_get_dst(struct thread_smc_args *args,
+	if (s) {
+		struct sp_shared_mem *ssm = NULL;
+
+		ssm = calloc(1, sizeof(struct sp_shared_mem));
+
+		if (!ssm)
+			return FFA_NO_MEMORY;
+
+		ssm->access_descr = mem_acc;
+
+		ssm->counter = 0;
+		ssm->endpoint_id = s->endpoint_id;
+
+		/* Only allow each endpoint ones */
+		if (get_sp_shared_mem_by_handle(s, global_handle)) {
+			free(ssm);
+			return FFA_DENIED;
+		}
+		/*
+		 * Make a link between struct shared_mem and
+		 * struct sp_shared_mem
+		 */
+		ssm->s_mem = mem_share;
+		SLIST_INSERT_HEAD(&s->mem_head, ssm, link);
+		SLIST_INSERT_HEAD(&mem_share->sp_head, ssm,
+				  link);
+
+	} else {
+		/*
+		 * We don't register memory that is shared with
+		 * the OP-TEE endpoint. However we do need to
+		 * check that if someone tries to share with an
+		 * invalid endpoint.
+		 */
+		if (endpoint_id != spmc_get_id())
+			return FFA_DENIED;
+	}
+
+	return FFA_OK;
+}
+
+TEE_Result spmc_sp_add_share(struct ffa_mem_transaction *input_descr,
+			     uint64_t global_handle, size_t blen)
+{
+	int rc = FFA_INVALID_PARAMETERS;
+	struct ts_session *ts = 0;
+	uint16_t caller_id = FFA_NW_ID;
+	unsigned int num_mem_accs = 0;
+	unsigned int i = 0;
+	void *mem_share_buffer = NULL;
+	struct ffa_mem_access *mem_acc = NULL;
+	struct shared_mem *mem_share = NULL;
+
+	ts = sp_get_active();
+
+	/* Find the endpoint which is sharing the memory region */
+	if (ts) {
+		struct sp_ctx *uctx = NULL;
+		struct sp_session *calling_s = NULL;
+
+		uctx = to_sp_ctx(ts->ctx);
+		calling_s = uctx->open_session;
+		caller_id = calling_s->endpoint_id;
+	}
+
+	num_mem_accs = READ_ONCE(input_descr->mem_access_count);
+
+	if (num_mem_accs) {
+		/*
+		 * We store the full incoming transaction buffer and have
+		 * shared_mem->mem_descr point to the beginning of the buffer.
+		 * A struct sp_shared_mem object is created for each
+		 * struct ffa_mem_access in the buffer.
+		 * sp_shared_mem->access_descr is set to point to the
+		 * corresponding struct ffa_mem_access in the transaction
+		 * buffer.
+		 */
+		mem_share = calloc(1, sizeof(struct shared_mem));
+		if (!mem_share)
+			return FFA_NO_MEMORY;
+
+		mem_share_buffer = calloc(1, blen);
+		if (!mem_share_buffer) {
+			free(mem_share);
+			return FFA_NO_MEMORY;
+		}
+
+		memcpy(mem_share_buffer, (void *)input_descr, blen);
+
+		/* Point mem_descr to the beginning of the buffer*/
+		mem_share->mem_descr = mem_share_buffer;
+		mem_share->mem_descr->global_handle = global_handle;
+		mem_share->owner_id = caller_id;
+
+		/* Create a list for all struct sp_shared_mem */
+		SLIST_INIT(&mem_share->sp_head);
+		TAILQ_INSERT_TAIL(&mem_shares, mem_share, link);
+
+		mem_acc = mem_share->mem_descr->mem_access_array;
+
+		/* Iterate over the mem_access_array */
+		for (i = 0; i < num_mem_accs; i++) {
+			rc = add_mem_region_to_sp(&mem_acc[i], mem_share,
+						  global_handle);
+			if (rc)
+				goto cleanup;
+		}
+		/* Return if we processed a valid message */
+		if (!rc)
+			return rc;
+	}
+
+cleanup:
+	if (num_mem_accs) {
```

Comment:
I'm not sure what's happening here. Are we undoing everything done above, regardless of outcome?